### PR TITLE
All should buld the docker images (and therefore the binaries and protos via dependancies), and not the exes directly.

### LINKED ISSUE #1371: Check in proto generated go files
Right now we do not check in any of the `*.pb.go` files that are generated.

During the [May 2nd Community Meeting](https://docs.google.com/document/d/1shtXSAqp3t7fiC-9uZcKkq3mgwsItAJlH6YW6x1joZo/edit#heading=h.ffnrx6acn0lc) we generally decided a good direction to go would be to:

1. Check in all generated code.
1. Pin to a specific gogo/protobuf version see #1370.
1. Create a CI job that fails if the checked in version does not match a newly generated version.

### PR DESCRIPTION
**Pull request by tomwilkie**
Also:
- reduce the spam when building by hiding the docker run commands.
- check in the protos so Cortex can be used as a dependancy (fixes #1371).
- [x] TODO: verify generated proto code in CI

On master, running `make` builds the binaries twice and enters the build container too often.

```
>>>> Entering build container: exes
make: Entering directory '/go/src/github.com/cortexproject/cortex'
protoc -I /go/src:./vendor:./pkg/chunk/storage --gogoslick_out=plugins=grpc:./pkg/chunk/storage ./pkg/chunk/storage/caching_index_client.proto
protoc -I /go/src:./vendor:./pkg/ingester/client --gogoslick_out=plugins=grpc:./pkg/ingester/client ./pkg/ingester/client/cortex.proto
protoc -I /go/src:./vendor:./pkg/querier/frontend --gogoslick_out=plugins=grpc:./pkg/querier/frontend ./pkg/querier/frontend/frontend.proto
protoc -I /go/src:./vendor:./pkg/distributor --gogoslick_out=plugins=grpc:./pkg/distributor ./pkg/distributor/ha_tracker.proto
protoc -I /go/src:./vendor:./pkg/ring --gogoslick_out=plugins=grpc:./pkg/ring ./pkg/ring/ring.proto
CGO_ENABLED=0 go build -ldflags "-extldflags \"-static\" -s -w" -tags netgo -o cmd/cortex/cortex ./cmd/cortex
CGO_ENABLED=0 go build -ldflags "-extldflags \"-static\" -s -w" -tags netgo -o cmd/test-exporter/test-exporter ./cmd/test-exporter
make: Leaving directory '/go/src/github.com/cortexproject/cortex'

real	2m31.642s
user	0m0.031s
sys	0m0.016s

>>>> Entering build container: generated
make: Entering directory '/go/src/github.com/cortexproject/cortex'
make: Nothing to be done for 'generated'.
make: Leaving directory '/go/src/github.com/cortexproject/cortex'

real	0m5.574s
user	0m0.033s
sys	0m0.018s

>>>> Entering build container: cmd/cortex/cortex
make: Entering directory '/go/src/github.com/cortexproject/cortex'
CGO_ENABLED=0 go build -ldflags "-extldflags \"-static\" -s -w" -tags netgo -o cmd/cortex/cortex ./cmd/cortex
make: Leaving directory '/go/src/github.com/cortexproject/cortex'

real	0m39.034s
user	0m0.027s
sys	0m0.017s

docker build --build-arg=revision=d0f8944cf7b03fc6a0c1f8a315ecffe28787c8e7 -t quay.io/cortexproject/cortex cmd/cortex/
...
```

Which this change we don't:

```
>>>> Entering build container: generated
make: Entering directory '/go/src/github.com/cortexproject/cortex'
protoc -I /go/src:./vendor:./pkg/chunk/storage --gogoslick_out=plugins=grpc:./pkg/chunk/storage ./pkg/chunk/storage/caching_index_client.proto
protoc -I /go/src:./vendor:./pkg/ingester/client --gogoslick_out=plugins=grpc:./pkg/ingester/client ./pkg/ingester/client/cortex.proto
protoc -I /go/src:./vendor:./pkg/querier/frontend --gogoslick_out=plugins=grpc:./pkg/querier/frontend ./pkg/querier/frontend/frontend.proto
protoc -I /go/src:./vendor:./pkg/distributor --gogoslick_out=plugins=grpc:./pkg/distributor ./pkg/distributor/ha_tracker.proto
protoc -I /go/src:./vendor:./pkg/ring --gogoslick_out=plugins=grpc:./pkg/ring ./pkg/ring/ring.proto
make: Leaving directory '/go/src/github.com/cortexproject/cortex'

real	0m12.988s
user	0m0.030s
sys	0m0.020s

>>>> Entering build container: cmd/cortex/cortex
make: Entering directory '/go/src/github.com/cortexproject/cortex'
CGO_ENABLED=0 go build -ldflags "-extldflags \"-static\" -s -w" -tags netgo -o cmd/cortex/cortex ./cmd/cortex
make: Leaving directory '/go/src/github.com/cortexproject/cortex'

real	1m56.313s
user	0m0.032s
sys	0m0.018s

docker build --build-arg=revision=d0f8944cf7b03fc6a0c1f8a315ecffe28787c8e7 -t quay.io/cortexproject/cortex cmd/cortex/
...
```

Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

**Comment by tomwilkie** [comment]
Remaining problem is that successive calls to `make` end up rebuilding when there are no changes.   I don't think this can be solved without reverting to building the protos separately.

File: Makefile [hunk 1]
```diff
@@ -9,10 +9,11 @@ GIT_REVISION := $(shell git rev-parse HEAD)
 UPTODATE := .uptodate
 
 # Building Docker images is now automated. The convention is every directory
-# with a Dockerfile in it builds an image calls quay.io/weaveworks/<dirname>.
+# with a Dockerfile in it builds an image calls quay.io/cortexproject/<dirname>.
 # Dependencies (i.e. things that go in the image) still need to be explicitly
 # declared.
 %/$(UPTODATE): %/Dockerfile
+	@echo
```
Comment (gouthamve) [inline Makefile]: What's this change for?

File: Makefile [hunk 2]
```diff
@@ -83,10 +82,12 @@ NETGO_CHECK = @strings $@ | grep cgo_stub\\\.go >/dev/null || { \
 
 ifeq ($(BUILD_IN_CONTAINER),true)
 
-exes $(EXES) generated $(PROTO_GOS) lint test shell mod-check: build-image/$(UPTODATE)
+exes $(EXES) protos $(PROTO_GOS) lint test shell mod-check check-protos: build-image/$(UPTODATE)
 	@mkdir -p $(shell pwd)/.pkg
 	@mkdir -p $(shell pwd)/.cache
-	$(SUDO) time docker run $(RM) $(TTY) -i \
+	@echo
```
Comment (gouthamve) [inline Makefile]: Same here.

File: Makefile [hunk 3]
```diff
@@ -96,7 +98,9 @@ configs-integration-test: build-image/$(UPTODATE)
 	@mkdir -p $(shell pwd)/.pkg
 	@mkdir -p $(shell pwd)/.cache
 	DB_CONTAINER="$$(docker run -d -e 'POSTGRES_DB=configs_test' postgres:9.6)"; \
-	$(SUDO) docker run $(RM) $(TTY) -i \
+	@echo
```
Comment (gouthamve) [inline Makefile]: Same here.

File: Makefile (hunk 1 above)
Comment (tomwilkie) [reply Makefile]: I put some new lines in the output of the makefile to make it easier to follow what it was doing.  I like them, but not super attached.  WDYT?

Comment (gouthamve) [reply Makefile]: Makes sense, thanks! LGTM!!!!

**Pull request merged**